src/main.go

```diff
@@ -484,27 +505,60 @@ func (t *tester) Run() error {
 		case Q_REMOVE_FILE:
 			err = os.Remove(strings.TrimSpace(q.Query))
 			if err != nil {
-				return errors.Annotate(err, "failed to remove file")
+				return testCnt, errors.Annotate(err, "failed to remove file")
 			}
 		case Q_REPLACE_REGEX:
 			t.replaceRegex = nil
 			regex, err := ParseReplaceRegex(q.Query)
 			if err != nil {
-				return errors.Annotate(err, fmt.Sprintf("Could not parse regex in --replace_regex: line: %d sql:%v", q.Line, q.Query))
+				return testCnt, errors.Annotate(
+					err, fmt.Sprintf("Could not parse regex in --replace_regex: line: %s sql:%v",
+						q.location(), q.Query))
 			}
 			t.replaceRegex = regex
-		default:
-			log.WithFields(log.Fields{"command": q.firstWord, "arguments": q.Query, "line": q.Line}).Warn("command not implemented")
-		}
-	}
+		case Q_SOURCE:
+			fileName := strings.TrimSpace(q.Query)
+			cwd, err := os.Getwd()
+			if err != nil {
+				return testCnt, err
+			}
 
-	fmt.Printf("%s: ok! %d test cases passed, take time %v s\n", t.testFileName(), testCnt, time.Since(startTime).Seconds())
+			// For security, don't allow to include files from other locations
+			fullpath, err := filepath.Abs(fileName)
+			if err != nil {
+				return testCnt, err
+			}
+			if !strings.HasPrefix(fullpath, cwd) {
+				return testCnt, errors.Errorf("included file %s is not prefixed with %s", fullpath, cwd)
+			}
 
-	if xmlPath != "" {
-		t.addSuccess(&testSuite, &startTime, testCnt)
-	}
+			// Make sure we have a useful error message if the file can't be found or isn't a regular file
+			s, err := os.Stat(fileName)
+			if err != nil {
+				return testCnt, errors.Annotate(err,
+					fmt.Sprintf("file sourced with --source doesn't exist: line %s, file: %s",
+						q.location(), fileName))
+			}
+			if !s.Mode().IsRegular() {
+				return testCnt, errors.Errorf("file sourced with --source isn't a regular file: line %s, file: %s",
+					q.location(), fileName)
+			}
 
-	return t.flushResult()
+			// Process the queries in the file
+			includedQueries, err := t.loadQueries(fileName)
+			if err != nil {
+				return testCnt, errors.Annotate(err, fmt.Sprintf("error loading queries from %s", fileName))
```

Comment:
Will this lead to a full 'stack trace'? I.e. so you can follow from the originating file's line, to the current `--source`'d file and line?
Is this testable? Maybe with a unit test?

Reply:
```
dvaneeden@dve-carbon:~/dev/pingcap/mysql-tester$ chmod 000 include/hello3.inc
dvaneeden@dve-carbon:~/dev/pingcap/mysql-tester$ ./mysql-tester source
ERRO[0000] run test [source] err: error loading queries from include/hello3.inc: open include/hello3.inc: permission denied 

ERRO[0000] 1 tests failed                               
ERRO[0000] run test [source] err: error loading queries from include/hello3.inc: open include/hello3.inc: permission denied 
```

Reply:
We could add the location:

```
dvaneeden@dve-carbon:~/dev/pingcap/mysql-tester$ ./mysql-tester source
ERRO[0000] run test [source] err: error loading queries from include/hello3.inc, line ./t/source.test:11: open include/hello3.inc: permission denied 

ERRO[0000] 1 tests failed                               
ERRO[0000] run test [source] err: error loading queries from include/hello3.inc, line ./t/source.test:11: open include/hello3.inc: permission denied 
dvaneeden@dve-carbon:~/dev/pingcap/mysql-tester$ git diff
diff --git a/src/main.go b/src/main.go
index 5e09363..3808f49 100644
--- a/src/main.go
+++ b/src/main.go
@@ -559,7 +559,7 @@ func (t *tester) runQueries(queries []query) (int, error) {
                        // Process the queries in the file
                        includedQueries, err := t.loadQueries(fileName)
                        if err != nil {
-                               return testCnt, errors.Annotate(err, fmt.Sprintf("error loading queries from %s", fileName))
+                               return testCnt, errors.Annotate(err, fmt.Sprintf("error loading queries from %s, line %s", fileName, q.location()))
                        }
                        includeCnt, err := t.runQueries(includedQueries)
                        if err != nil {
```